pop_art_agent.py

```diff
@@ -137,32 +139,53 @@ def interact(self):
         done, prv_obs, prv_core_state = self._retrive_values()
         initial_core_state = prv_core_state
 
+        s0 = self.state_encoder.initial_state(
+            batch_size=self.num_env)
+        # FIXME(ycho):
+        # I really don't like these tuple-based
+        # state representations...
+        s0 = tuple(s.to(self.device) for s in s0)
+
         for _ in range(self.num_interactions):
             with th.no_grad():
                 state, core_state = self.state_encoder(
-                    prv_obs, prv_core_state, th.tensor(done, dtype=th.bool).unsqueeze(dim=0)
-                )
+                    prv_obs, prv_core_state, th.as_tensor(
+                        done, dtype=th.bool, device=self.device)[None])
                 dist = self.get_action_distribution(state)
                 action = dist.sample().squeeze()
                 # NOTE(ycho): store `log_prob` for vtrace calculation
                 log_prob = dist.log_prob(action).squeeze()
 
-            obs, rew, done, info = self.env.step(action.cpu().numpy())
+            obs, rew, done, info = self.env.step(action.detach().cpu().numpy())
             obs = self._format_observations(obs)
             rew = np.asarray(rew, dtype=np.float32)
             buf.append((prv_obs, action, log_prob, obs, rew, done))
             prv_obs = obs
-            prv_core_state = core_state
 
-        self._save_values(done, prv_obs, core_state)
+            # NOTE(ycho): Need to reset initial states
+            # for all environments that have terminated.
+            # Since this operation might not be needed,
+            if np.any(done):
+                prv_core_state = tuple(
+                    th.where(
+                        th.as_tensor(
+                            done[None, :, None],
+                            device=s0[i].device),
+                        s0[i],
+                        core_state[i]) for i in range(2))
```

Comment:
This might not be necessary, since the network resets rnn state of the environments that have terminated, if my understanding is correct. Anyway, I think it is okay (or safe) to leave it this way.

Reply:
ok